sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SerializationConstructorsShouldBeSecured.cs

```diff
@@ -0,0 +1,139 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class SerializationConstructorsShouldBeSecured : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4212";
+        private const string MessageFormat = "Secure that serialization constructor.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly AttributeComparer attributeComparer = new AttributeComparer();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(FindPossibleViolations,
+                SyntaxKind.ConstructorDeclaration);
+        }
+
+        private void FindPossibleViolations(SyntaxNodeAnalysisContext c)
+        {
+            var constructorSyntax = (ConstructorDeclarationSyntax)c.Node;
+            var reportLocation = constructorSyntax?.Identifier.GetLocation();
+            if (reportLocation == null)
+            {
+                return;
+            }
+
+            var serializationConstructor = c.SemanticModel.GetDeclaredSymbol(constructorSyntax);
+            if (!serializationConstructor.IsSerializationConstructor())
+            {
+                return;
+            }
+
+            var classSymbol = serializationConstructor.ContainingType;
+            if (!classSymbol.Implements(KnownType.System_Runtime_Serialization_ISerializable))
+            {
+                return;
+            }
+
+            var isAssemblyIsPartiallyTrusted = c.SemanticModel.Compilation.Assembly
+                    .GetAttributes()
+                    .Any(a => a.AttributeClass.Is(KnownType.System_Security_AllowPartiallyTrustedCallersAttribute));
+            if (!isAssemblyIsPartiallyTrusted)
+            {
+                return;
+            }
+
+            var serializationConstructorAttributes = GetCasAttributes(serializationConstructor);
+
+            bool isConstructorMissingAttributes = classSymbol.Constructors
+                .SelectMany(m => GetCasAttributes(m))
+                .Any(attr => !serializationConstructorAttributes.Contains(attr, attributeComparer));
+
+            if (isConstructorMissingAttributes)
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, reportLocation));
+            }
+        }
+
+        private static IEnumerable<AttributeData> GetCasAttributes(IMethodSymbol methodSymbol)
+        {
+            return methodSymbol.GetAttributes().Where(IsCasAttribute);
+
+            bool IsCasAttribute(AttributeData data) =>
```

Comment:
Please use a long name. I had to look at the full name to understand this method.